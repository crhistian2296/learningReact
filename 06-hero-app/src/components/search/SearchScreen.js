import queryString from 'query-string';
import { useMemo } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import useForm from '../../hooks/useForm';
import { getHeroByName } from '../../selectors/getHeroByName';
import { HeroCard } from '../heroes/HeroCard';

const SearchScreen = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { q = '' } = queryString.parse(location.search);

  const [formValues, handleInputChange] = useForm({
    searchText: q,
  });

  const { searchText } = formValues;
  const herosFiltered = useMemo(() => getHeroByName(q), [q]);

  const handleSearch = (e) => {
    e.preventDefault();
    navigate(`?q=${searchText}`);
  };

  return (
    <>
      <div className='row'>
        <div className='col-sm-5 mb-4'>
          <h3>Search</h3>
          <hr />

          <form onSubmit={handleSearch}>
            <input
              type='text'
              className='form-control'
              name='searchText'
              placeholder='Ej: Flash'
              autoComplete='off'
              onChange={handleInputChange}
              value={searchText}
            />

            <button type='submit' className='btn btn-outline-primary mt-2'>
              Enviar
            </button>
          </form>
        </div>
        <div className='col-sm-7'>
          <h3>Results</h3>
          <hr />

          {q === '' ? (
            <div className='alert alert-info animate__animated animate__fadeIn'>
              Search your hero
            </div>
          ) : (
            herosFiltered.length === 0 && (
              <div className='alert alert-danger animate__animated animate__fadeIn'>{`There is no results for: ${q}`}</div>
            )
          )}

          {herosFiltered &&
            herosFiltered.map((hero) => (
              <>
                <HeroCard key={hero.id} {...hero} className='mb-3' />
                <div className='mb-3'></div>
              </>
            ))}
        </div>
      </div>
    </>
  );
};
export default SearchScreen;
