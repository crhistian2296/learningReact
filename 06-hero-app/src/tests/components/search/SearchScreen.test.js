import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import SearchScreen from '../../../components/search/SearchScreen';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('Pruebas en <SearchScreen/>', () => {
  test('Debe de hacer el render con valores por defecto', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info').text().trim()).toBe('Search your hero');
  });

  test('Debe de mostrar a Batman y el input con el queryString', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchScreen />
      </MemoryRouter>
    );
    // const queryParameter = window.location.pathname.trim();
    // console.log(queryParameter);

    expect(wrapper.find('h5').text().trim()).toBe('Batman');
    expect(wrapper.find('input').prop('value').trim()).toBe('batman');
  });

  test('Debe de mostrar alerta de busqueda fallida', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper.find('.alert-danger').text().trim()).toBe('There is no results for: batman123');
  });

  test('Debe llamar el navigate a la nueva pantalla', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    );

    wrapper.find('input').simulate('change', { target: { name: 'searchText', value: 'batman' } });

    // const useNavigate = jest.fn();

    wrapper.find('form').prop('onSubmit')({ preventDefault() {} });
    expect(mockNavigate).toHaveBeenCalledWith('?q=batman');
  });
});
