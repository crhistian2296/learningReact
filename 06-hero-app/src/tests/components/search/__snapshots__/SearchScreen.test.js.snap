// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen/> Debe de hacer el render con valores por defecto 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-5 mb-4"
  >
    <h3>
      Search
    </h3>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="Ej: Flash"
        type="text"
        value=""
      />
      <button
        className="btn btn-outline-primary mt-2"
        type="submit"
      >
        Enviar
      </button>
    </form>
  </div>
  <div
    className="col-sm-7"
  >
    <h3>
      Results
    </h3>
    <hr />
    <div
      className="alert alert-info animate__animated animate__fadeIn"
    >
      Search your hero
    </div>
  </div>
</div>
`;
